[step1]: 문자열 계산기 과제

RoadtoS7 opened the pull request:
문자열 계산기 과제 제출합니다~!

사칙연산시에 아래와 같이 when을 사용해서 연산자에 따라 연산을 수행하도록 했습니다.

```kotlin
 fun checkThenCalculate(operator: String, result: Double, number: String): Double {
        checkNumber(number)
        checkOperator(operator)
        return when (operator) {
            "+" -> add(result, number.toDouble())
            "-" -> subtract(result, number.toDouble())
            "*" -> multiply(result, number.toDouble())
            "/" -> divide(result, number.toDouble())
            else -> throw IllegalArgumentException("연산자 위치에 잘못된 값이 들어있습니다. 잘못된 연산자: $operator")
        }
    }
```

이때 연산자 기호를 "+", "-", "*", "/" 이렇게 하드 코딩으로 사용했는데
객체지향생활체조원칙에서  "모든 원시 값과 문자열을 포장한다." 라는 말에 따르면 이런 것들을 모두 변수에 담아서 사용해야 한다는 것인지 궁금합니다!

그리고 우선은 테스트를 모든 함수마다 해보고 싶어서 모두 public으로 선언했는데 
private함수를 사용할경우 public 함수를 통해서 private 함수를 사용하라고 저번 스터디시간에 말씀해주셨습니다.
하지만 어디까지 private으로 묶어야 할지가 아직까지 감이 잡히지 않습니다..ㅠㅠㅠ 
(하다보면 늘겠죠...?ㅜㅜ)

RoadtoS7 committed:
feat(Calculator.kt): create Calculator.kt

연산을 수행하는 계산기 클래스를 만들었다.

RoadtoS7 committed:
test(Calculator.kt): add test code for Calculator class

계산기 클래스를 수정하는 테스트 코드를 만들었다.

RoadtoS7 committed:
feat(Main.kt): add driver file which have main function

사용자로부터 입력을 받아서 계산을 수행하는 메인함수를 담고 있는 Main.kt 파일을 만들었다.

malibinYun commented on src/main/kotlin/stringcaclulator/Calculator.kt:
```diff
@@ -0,0 +1,76 @@
+import java.lang.IllegalArgumentException
+import kotlin.math.round
+
+class Calculator {
```
계산기 객체가 너무 많은 일을 하고 있다고 생각해.
 1. 쪼개는 기능
 2. 검증하는 기능
 3. 계산하는 기능
 ...

클래스가 이 일까지 해야 할 책임이 있을까? 하고 되물어보는 게 좋을거 같아.

더 힌트를 주자면, "계산기"가 "쪼깨는"일을 할 책임을 가지고있을까?

RoadtoS7 replied:
오옹 힌트 감사합니다! 클래스 분리해서 다시 해보겠습니다!!

malibinYun commented on src/main/kotlin/stringcaclulator/Calculator.kt:
```diff
@@ -0,0 +1,76 @@
+        return calculateWithSplitList(inputList)
+    }
+
+    fun splitInput(input: String): List<String> {
+        return input
+            .split(" ")
+    }
```
return 은 인라인으로 해도 될거같아.

malibinYun replied:
그보다, splitInput(list)와 그냥 list.split(" ") 중에서 나는 후자가 더 가독성이 좋아보여.
굳이 래핑하지 않아도 될 메서드같아!

malibinYun replied:
또. input은 되게 추상레벨이 높아서, 나라면 stringTokens 내지는 tokens 이런 이름을 썼을거 같아.

RoadtoS7 replied:
생각해보니까 함수로 만들지 않는게 나을 것 같네요! 이 부분도 반영하겠습니다~

malibinYun commented on src/main/kotlin/stringcaclulator/Calculator.kt:
```diff
@@ -0,0 +1,76 @@
+            .split(" ")
+    }
+
+    fun calculateWithSplitList(splitList: List<String>): Double {
```
메서드 명이 자세하면 좋긴한데! 적당한 추상화가 더 좋다고 처음 카톡방에 공유해준 객체지향 글에 적혀있더라구.

예를들면, 
"냄비에 물을 넣고 불을 킨 화구에 올리기"
보다
"물끓이기"
가 더 좋은 표현인거지!

RoadtoS7 replied:
함수 이름에 대해서 한번더 고민해보도록 하겠습니다~!

malibinYun commented on src/main/kotlin/stringcaclulator/Calculator.kt:
```diff
@@ -0,0 +1,76 @@
+        return result
+    }
+
+    fun checkThenCalculate(operator: String, result: Double, number: String): Double {
+        checkNumber(number)
+        checkOperator(operator)
+        return when (operator) {
+            "+" -> add(result, number.toDouble())
+            "-" -> subtract(result, number.toDouble())
+            "*" -> multiply(result, number.toDouble())
+            "/" -> divide(result, number.toDouble())
+            else -> throw IllegalArgumentException("연산자 위치에 잘못된 값이 들어있습니다. 잘못된 연산자: $operator")
+        }
+    }
```
파라미터 3개 !!! 2개 이하로 줄여보자!!
이 메서드도 두가지의 일을 하고있다고 생각해. validation로직과 계산 로직. 한 번 분리해보는게 어떨까?

malibinYun commented on src/main/kotlin/stringcaclulator/Calculator.kt:
```diff
@@ -0,0 +1,76 @@
+        }
+    }
+
+    fun checkNumber(number: String): Boolean {
+        val result = "[0-9]+|[0-9]+[.][0-9]+".toRegex()
+            .matches(number)
+        if (!result) {
+            throw IllegalArgumentException("숫자 위치에 잘못된 값이 들어있습니다. 잘못된 숫자: $number")
+        }
+        return result
+    }
+
```
자바에서는 정규표현식 객체를 만들때 비용이 매우 크다고 알고있어. 코틀린이라고 다르지 않을꺼야.
그렇다면, 이 메서드를 호출할 때마다 비용을 감당해야할까?

malibinYun replied:
"~~~" 같은 변하지 않을 상수는 코드에 하드코딩 하는 것 보다 아예 상수필드로 빼는게 좋아. 다른 사람들이 봤을 때 저 정규식이 조금만 복잡해져도 뭘 찾고싶어하는지 절대로 알 수 없을거라서 이름도 부여할 수 있고 좋지.
상수로 빠지는 Activity.RESULT_OK 처럼 상수로 빼보자.

malibinYun replied:
예외를 터트릴 때 잘못된 게 무엇인지 알려주는거 완전 잘했어 !! 굳굳👍👍

RoadtoS7 replied:
제가 이해한 것이 맞는지 한번더 여쭙습니다! 정규 표현식을 클래스의 멤버 변수이면서 상수로 만드는 것이 좋다는 말씀이시죠??

malibinYun replied:
음 멤버변수라고 말을하던가.... 어쨋든 의미는 비슷해 보입니다~~

RoadtoS7 replied:
감사합니다!!

malibinYun commented on src/main/kotlin/stringcaclulator/Calculator.kt:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    fun checkNumber(number: String): Boolean {
+        val result = "[0-9]+|[0-9]+[.][0-9]+".toRegex()
```
자바에서 Boolean 값에 해당하는 변수는 is~~~ 처럼 관용적으로 사용되고 있어.
```suggestion
        val isNumber = ~~~
```

RoadtoS7 replied:
오 감사합니다! 이 부분도 반영할게요!!

malibinYun commented on src/main/kotlin/stringcaclulator/Calculator.kt:
```diff
@@ -0,0 +1,76 @@
+        return result
+    }
+
+    fun <T : Number> add(number1: T, number2: T) =
+        number1.toDouble() + number2.toDouble()
```
제네릭을 바른 이유가 뭔지 궁금해! ?.?

RoadtoS7 replied:
처음에는 사칙연산을 하는 함수에 파라미터로 Int 타입의 데이터도 들어갈 수 있고, Double 타입의 데이터도 들어갈 수 있다고 생각해서 둘다를 받을 수 있도록 함수를 만들기 위해서 제네릭으로 했는데요..! 
생각해보니까 문자열로 입력을 받기 때문에 연산을 하려면 double로 바꿔주니까 제네릭으로 사용할 필요가 없을것 같아요!
수정하겠습니다~!

malibinYun commented on src/test/kotlin/stringcalculator/TestCalculator.kt:
```diff
@@ -0,0 +1,79 @@
+        assertThat(result).isEqualTo(true)
+    }
+
+    @Test
+    fun checkOperatorTest() {
+        val calculator = Calculator()
+        val result = calculator.checkOperator("-")
+        assertThat(result).isEqualTo(true)
+    }
```
validation 로직의 경우에는 리턴이 없는 경우가 많아.
왜냐면 잘못되면 에러가 throw되어서 어플리케이션이 터지기 때문에 리턴값이 필요가없어.
그러니 exception을 활용한 테스트로 써보자!

RoadtoS7 replied:
넵 반영하겠습니다!

malibinYun commented on src/test/kotlin/stringcalculator/TestCalculator.kt:
```diff
@@ -0,0 +1,79 @@
+import stringcaclulator.Calculator
+
+class TestCalculator {
+    @Test
+    fun calculateTest() {
+        val calculator = Calculator()
+        val result = calculator.calculate("3.0 - 4.5 * 2.7 + 11.05")
+        assertThat(result).isEqualTo(((3.0 - 4.5) * 2.7) + 11.05)
+    }
```
테스트 시나리오가 굳굳 👍👍
딱 봐도 calculator의 calculate가 뭐하는 기능인지 알 수 있었어.

RoadtoS7 replied:
섬세한 피드백 감사합니다!! 
덕분에 많이 배우고 있어요ㅎㅎ

malibinYun reviewed (changes requested):
과제 제출 고생했어 !! 피드백 반영해서 리팩터링 진행해보자~~ 😀😀

when 써서 가독성이 높은 코드가 된거같아 ㅎㅎ 스터디 때 좀 더 가독성 좋은 방법에 대해 알려줄게.
객체지향 생활체조 원칙보다는 지금은 내가 따로 프로그래밍 요구사항을 정리해둔 거에 초점을 맞춰서 해주면 좋을거같아. 아직 원시 값 포장에 대해서 가르쳐주지 않았으니까!

private, public의 경우는 해당 클래스에서 공개하고 싶은 메서드는 public, 공개할 필요가 없고 감춰야할 애들은 private로 두면 돼. 예를들면... 클래스 내부에서만 쓰는 메서드를 굳이 public으로 둘 필요가 없달까?

아래 예외사항들이 정확히 걸러지지 않고있어.
* "2 3", "2 *" 입력 시 IndexOutOfBound Exception 발생
* 공백 입력 시 숫자 위치에 잘못 된 값 들어있다는 Exception 발생 (기댓값 : 0 또는 공백이 아닌 수식을 입력하라는 Exception)

Caculator 클래스를 만든건 잘했어! 하지만 이 클래스가 God클래스가 된 느낌이야. 모든 일을 다 혼자서 할 수 있는 능력을 가진 클래스. 객체 지향은 각 책임을 가진 객체들이 협력해서 이루어 지는게 목표니까, 책임을 분리해보자!

engineer-choi reviewed (commented):
고생 많아유 수현~

jinsu4755 reviewed (commented):
혁이형이 너무 잘 집어주셔서 저도 비슷하게 궁금해서 더 남길 코맨트가 없습니다~ 고생하셨습니다~

RoadtoS7 committed:
refactor(Calculator.kt): remove unnecessary generic key word

사칙연산을 수행하는 함수에서 필요없는 제네릭 키워드를 삭제했습니다.

RoadtoS7 committed:
refactor(Calculator.kt): make regular expression for number and operator to be instance variable

정규표현식을 만들 때의 비용이 많이 들어가므로
숫자와 연산자의 유효성을 체크하기 위해서 사용되는 정규표현식을 정규표현식 객체로
만들어서 static으로 선언해 놓았다.

RoadtoS7 committed:
refactor(Calculator.kt): remove splitInput to make code more readable

코드의 가독성을 높이기 위해서 splitInput함수를 만들지 않고,
그대로 사용하도록 만들었다.

RoadtoS7 committed:
feat(Checker.kt): add checkMathExpressionSize function to check mathExpressionSize

입력 요소의 개수가 최소 3개는 되어야 수식이 되기 때문에 이를 확인하는 함수인 checkMathExpressionSize()를 만들었다.

RoadtoS7 committed:
refactor(Calculator.kt): apply code review

수식, 숫자, 연산자의 유효성을 체크하는 Checker 클래스 객체를 인스턴스 변수로
가지고 이를 이용하여 검증하는 기능을 수행하도록 했다.
checkThenCalculate()가 검증하는 기능과 계산하는 기능 두가지를 수행하여,
계산하는 기능만 수행하도록 하고 함수의 이름을 calculateOperator()로
바꾸었다.
checkThenCalculate()의 인자가 3개였는데 이를 줄이기 위해서 결과값을 담는
변수를 클래스내에 result라는 이름의 인스턴스 변수로 만들었다.

RoadtoS7 committed:
test(TestCalculator.kt): apply code review

checker객체에 새롭게 만든 함수인 checkMathExpressionSize를 검사하는
테스트 코드(mathExpressionTest)를 만들었다.
숫자, 연산자의 유효성을 검사하는 함수에서 반환값을 없애고
assertThatThrownBy()함수를 이용하여 exception을 활용한 테스트를
하도록 만들었다.

RoadtoS7 commented:
궁금한게 생겼습니다!
테스트 코드를 수행하는 함수를 작성할 때, 보통 성공케이스에 대한 함수와 실패케이스에 대한 함수를 둘다 작성하나요??

RoadtoS7 committed:
refactor(Checker.kt): 검사를 수행하는 클래스를 object로 만듬

Checker 클래스의 경우 util함수와 상수만을 가지고 있으므로 object로 만들어도 된다고 판단했으며,
Calculator 클래스에서 Checker의 함수를 사용하기 위해 Checker객체를
멤버 변수로 가지고 있는 것보다 object로 만들어서 직접 접근하는 것이 의존성을
두 클래스간의 낮출 수 있다는 생각이 들었다.

RoadtoS7 committed:
test(CheckerTest.kt):  add test code file for Checker class

Checker 클래스 내의 함수를 테스트 하기 위한 테스트 코드를 만들었다.
각 함수별로 옳바른 테스트 케이스와 잘못된 것을 판단하기 위한 테스트를 만들었다.
입력을 리스트형태로 제공하기 위해서 generateWrongMathExpression()함수와
generateMathExpression()함수를 만들어 이용했다.

RoadtoS7 committed:
refactor(Calculator.kt): add enum class for operator

연산자를 상수처리하기 위해서 enum class를 만들었다.
그리고 enum class내에 연산자에 해당하는 객체가 존재하는지 확인하기 위한 함수를
만들었다.

RoadtoS7 committed:
test(CalculatorTest.kt): add parameter to each test function

ParameterizedTest 어노테이션을 이용하여 각 테스트 함수가 파라미터를 받아서
동작할 수 있도록 만들었다.
Calculator 사용자의 문자열 입력값을 받아서 계산을 시작하는 함수인
calculate()를 테스트하기 위해서 MethodSource 어노테이션과
generateCalculateParameter()함수를 이용했다.
왜냐하면 입력값으로 연산식을 넣어서 연산식의 결과값이 파라미터로 전달될 수 있도록
하고 싶었는데, 이렇게 타입이 다르거나 리스트와 같이 복잡한 입력은 MethodSource
어노테이션을 사용해야만 가능했다.

RoadtoS7 commented:
2주차 과제 제출합니다!
스트레터지 패턴을 Operator 이넘 클래스에 적용하고 싶었으나
연산을 수행하는 함수는 계산하는 역할인 Calculator 클래스에서 갖고 있는 것이 맞다고 판단하여 적용하지 않았습니다. 
Calculator 클래스나 Checker클래스에 적용하는 방향도 생각했지만, 
둘다 여러개의 객체를 생성해서 각 객체마다 수행하는 전략이 다른 것은 아니라고 생각했습니다.
저는 이렇게 생각했지만, 문자열 계산기에 전략 패턴을 적용하면 좋은 경우가 있다면 알고싶습니다!

malibinYun commented on src/main/kotlin/stringcaclulator/Calculator.kt:
```diff
@@ -0,0 +1,78 @@
+    MULTIPLY("*"),
+    DIVIDE("/");
+
+    fun hasSymbol(otherSymbol: String) =
+        this.operator == otherSymbol
```
```suggestion
    fun hasSymbol(otherSymbol: String) = this.operator == otherSymbol
```
한 줄 짜리면 줄바꿈 안해도 돼!

malibinYun commented on src/main/kotlin/stringcaclulator/Checker.kt:
```diff
@@ -0,0 +1,26 @@
+    private val NUMBER_REGEX = "[0-9]+|[0-9]+[.][0-9]+".toRegex()
+    private val OPERATOR_REGEX = "[+]|[-]|[*]|[/]".toRegex()
+
+    fun checkMathExpressionSize(mathExpression: List<String>) {
+        require(mathExpression.size >= 3) { throw IllegalArgumentException("연산식의 구성요소는 최소 3개여야 합니다.") }
+    }
```
인풋이 0 하나면 0이 아웃풋으로 나와야하지 않을까?

malibinYun commented on src/main/kotlin/stringcaclulator/Checker.kt:
```diff
@@ -0,0 +1,26 @@
+import java.lang.IllegalArgumentException
+
+object Checker {
+    private val NUMBER_REGEX = "[0-9]+|[0-9]+[.][0-9]+".toRegex()
+    private val OPERATOR_REGEX = "[+]|[-]|[*]|[/]".toRegex()
```
캡슐화 👍👍

malibinYun commented on src/main/kotlin/stringcaclulator/Checker.kt:
```diff
@@ -0,0 +1,26 @@
+        require(mathExpression.size >= 3) { throw IllegalArgumentException("연산식의 구성요소는 최소 3개여야 합니다.") }
+    }
+
+    fun checkNumber(number: String) {
+        val isNumber = NUMBER_REGEX.matches(number)
+        if (!isNumber) {
+            throw IllegalArgumentException("숫자 위치에 잘못된 값이 들어있습니다. 잘못된 숫자: $number")
+        }
+    }
```
```suggestion
    fun checkNumber(number: String) {
        if (!NUMBER_REGEX.matches(number)) {
            throw IllegalArgumentException("숫자 위치에 잘못된 값이 들어있습니다. 잘못된 숫자: $number")
        }
    }
```
요건 인라인도 가독성이 나쁘지 않아보여! 
그리구 require 사용을 습관화해보자!

malibinYun commented on src/main/kotlin/stringcaclulator/Calculator.kt:
```diff
@@ -0,0 +1,78 @@
+        }
+    }
+
+    private fun add(number1: Double, number2: Double): Double {
+        val result = number1 + number2
+        return round(result * 1000) / 1000.0
+    }
+
+    private fun subtract(number1: Double, number2: Double): Double {
+        val result = number1 - number2
+        return round(result * 1000) / 1000.0
+    }
+
+    private fun multiply(number1: Double, number2: Double) = number1 * number2
+
+    private fun divide(number1: Double, number2: Double) = number1 / number2
+}
```
음 이거는 전략패턴을 쓰지 않아도 여전히 불필요한 메서드 인거같아.
Operator.PLUS -> acc + number
이런식으로?

malibinYun commented on src/test/kotlin/stringcaclulator/CalculatorTest.kt:
```diff
@@ -0,0 +1,91 @@
+        assertThat(result).isEqualTo(expectedResult)
+    }
+
+    @DisplayName("더하기 연산 테스트")
+    @ParameterizedTest
+    @CsvSource(
+        "0.57 + 0.48, 1.05",
+        "3 + 4, 7",
+        "1.00 + 5, 6.0",
+    )
+    fun addTest(mathExpression: String, expectedResult: Double) {
+        // given
+        val calculator = Calculator()
+        // when
+        val result = calculator.calculate(mathExpression)
+        assertThat(result).isEqualTo(expectedResult)
+    }
```
더하기에 대한 여러 개를 테스트해본건 좋은 접근이야!
하지만 "더한다" 말고 다른 갈래가 없는 기능은 1개만 테스트를 해줘도 좋아.
"계산한다"는 4개의 갈래가 나뉘니 그걸 합쳐서 1개의 테스트로 4개의 테스트를 진행하는게 더 좋을거같아.

malibinYun commented on src/test/kotlin/stringcaclulator/CheckerTest.kt:
```diff
@@ -0,0 +1,75 @@
+            .isInstanceOf(IllegalArgumentException::class.java)
+    }
+
+    @DisplayName("연산자 유효성 테스트")
+    @ParameterizedTest
+    @CsvSource("+", "-", "*", "/")
+    fun checkOperator(operator: String) {
+        assertThatCode { Checker.checkOperator(operator) }
+            .doesNotThrowAnyException()
+    }
```
정상인 것들을 정상인지 테스트하는 것 보다, 코드에는 비정상 인게 들어올 경우 에러가 나오는걸 의도했으니 비정상에 대한 것을 테스트하는게 좋아!
밑에 메서드가 있으니 이건 없어도 좋을거같구, 또, .hasMessage()로 어떤 에러메시지가 나왔는지 체크 해주는것도 중요해!

malibinYun commented on src/test/kotlin/stringcalculator/TestCalculator.kt:
```diff
@@ -0,0 +1,78 @@
+import stringcaclulator.Checker
+import java.lang.IllegalArgumentException
+
+class TestCalculator {
+    @Test
+    fun calculateTest() {
```
필요없어진 클래스는 삭제!!

malibinYun reviewed (approved):
전반적으로 많이 깔끔해졌네!! 아쉬운 점이 아직 있지만 계속 하다보면 더 좋아질거야
다음 리팩터링을 더 하고싶으면 머지해달라구 말해줘. 우선은 머지안하고 열어둘게!

malibinYun commented:
전략패턴을 쓰지 않은건 조금 의외네. 내 생각에는 Operator 라는 객체가 생김으로써 Calculator에서 연산기호에 해당하는 계산 방법을 선택하는 것은 결합도가 증가하는 것으로 보여. Operator가 없을 땐 맞는 책임이었지만 말이야. Operator가 계산 방법에 대해서 가지고있는 게 응집도가 높고 더 객체지향적인 코드라고 생각해. 실제 계산하는 게 아닌, "계산 방법"에 대해 가지고 있는거니까!

RoadtoS7 committed:
test(TestCalculator.kt): delete TestCalculator.kt

RoadtoS7 committed:
refactor(Calculator.kt): Refactor enum class Operator to use strategy pattern

Operator 이넘 클래스가 전략 패턴을 사용하도록 수정했다.

RoadtoS7 committed:
test(CalculatorTest.kt): refactor CalculatorTest.kt

RoadtoS7 committed:
Refactor(Main.kt): refactor Main.kt
